### PR TITLE
EMSUSD-1518 As an artist, I want the ability to determine whether a collection includes all objects by default or not

### PR DESCRIPTION
**Pull request by barbalt**
Initial iteration of the new light attributes:
* Added a new attribute to Usd Lights
* Lists of included and excluded prims to the light collection
* Ability to "Include all" which sets the `includeRoot` light attribute

The main new file is the `widget.py` under the collection folder. The common folder are for inheriting different Qt classes and themes.

image

### CODE REVIEW

File: lib/mayaUsd/resources/ae/usd-shared-components/src/python/usdSharedComponents/common/theme.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            return value * self.uiScaleFactor
+        if isinstance(value, int):
+            return int(round(float(value) * self.uiScaleFactor))
+        raise ValueError("Value must be a float or an int")
+
+    def uiUnScaled(self, value: float | int):
+        ### Returns the scaled value.
+        return float(value) / self.uiScaleFactor
+
+    def icon(self, name: str) -> QIcon:
```
Comment (AramAzhari-adsk) [inline lib/mayaUsd/resources/ae/usd-shared-components/src/python/usdSharedComponents/common/theme.py:50]: Is this being used somewhere?

Comment (barbalt) [reply lib/mayaUsd/resources/ae/usd-shared-components/src/python/usdSharedComponents/common/theme.py:50]: No, I can remove the icon function

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 2]
```diff
@@ -204,6 +205,55 @@ def get(self):
...
+                ptr = MQtUtil.findControl(self.parent)
+                parentWidget = wrapInstance(int(ptr), QWidget)
+
+                from usd_shared_components.collection.widget import CollectionWidget
+
+                self.widget = CollectionWidget(Usd.CollectionAPI.Get(self.prim, 'lightLink'))
+                parentWidget.layout().addWidget(self.widget)
+                
+            except Exception as ex:
+                logger.exception(ex)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:245]: Why do we need a logger? Can you just use `print` instead? Maya captures stdout for the Script Editor.

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 3]
```diff
@@ -204,6 +205,55 @@ def get(self):
             folderIndex[groups].items.append(attributeInfo.name)
         return self._attributeLayout
 
+import logging
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+class LightLinkingCustomControl(object):
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:212]: Please place this in a separate file like all the other custom controls.

**Review by seando-adsk — CHANGES_REQUESTED** [review]
custom control should be a separate file.

File: lib/mayaUsd/resources/ae/usdschemabase/lightCustomControl.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                ptr = MQtUtil.findControl(self.parent)
+                parentWidget = wrapInstance(int(ptr), QWidget)
+
+                from usd_shared_components.collection.widget import CollectionWidget # type: ignore
+
+                self.widget = CollectionWidget(Usd.CollectionAPI.Get(self.prim, 'lightLink'))
+                parentWidget.layout().addWidget(self.widget)
+                
+            except Exception as ex:
+                print(ex)
```
Comment (AramAzhari-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/lightCustomControl.py:37]: Not a huge deal but could be a bit more descriptive:
`print('Failed to create Light custom control: %s' % (ex))`

**Review by AramAzhari-adsk — DISMISSED** [review]

File: lib/mayaUsd/resources/ae/CMakeLists.txt [hunk 5]
```diff
@@ -37,3 +38,21 @@ foreach(_SUBDIR ${MAYAUSD_AE_TEMPLATES})
...
+)
+
+install(FILES
+    ${MAYAUSD_SHARED_COMPONENTS}/common/__init__.py
+    ${MAYAUSD_SHARED_COMPONENTS}/common/list.py
+    ${MAYAUSD_SHARED_COMPONENTS}/common/persistentStorage.py
+    ${MAYAUSD_SHARED_COMPONENTS}/common/resizable.py
+    ${MAYAUSD_SHARED_COMPONENTS}/common/theme.py
+    DESTINATION ${CMAKE_INSTALL_PREFIX}/lib/python/usd_shared_components/common/
+)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/CMakeLists.txt:58]: As agreed, we won't support this for 2022, so just wrap all this in `if(MAYA_APP_VERSION VERSION_GREATER_EQUAL 2023)`

File: lib/mayaUsd/resources/ae/CMakeLists.txt [hunk 6]
```diff
@@ -29,6 +29,7 @@ foreach(_SUBDIR ${MAYAUSD_AE_TEMPLATES})
         ${_SUBDIR}/displayCustomControl.py
         ${_SUBDIR}/enumCustomControl.py
         ${_SUBDIR}/imageCustomControl.py
+        ${_SUBDIR}/lightCustomControl.py
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/CMakeLists.txt:32]: Move this one to the section below and just install it only for 2023.

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 7]
```diff
@@ -21,8 +21,10 @@
 from .displayCustomControl import DisplayCustomControl
 from .materialCustomControl import MaterialCustomControl
 from .metadataCustomControl import MetadataCustomControl
+from .lightCustomControl import LightLinkingCustomControl
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:24]: I think you could wrap this in a try/except block and set a boolean variable such as `lightLinkingSupported`. Then at line 649 below only do it when that variable is True.

**Review by seando-adsk — CHANGES_REQUESTED** [review]
As agreed we won't support this in 2022.

**Review by AramAzhari-adsk — APPROVED** [review]

**Pull request merged**